sematic/ui/src/hooks/logHooks.ts

```diff
@@ -0,0 +1,110 @@
+import { useCallback, useEffect, useMemo, useState } from "react";
+import useAsyncFn from "react-use/lib/useAsyncFn";
+import useList from "react-use/lib/useList";
+import useLatest from "react-use/lib/useLatest";
+import { LogLineRequestResponse } from "../Payloads";
+import { useLogger } from "../utils";
+import { useHttpClient } from "./httpHooks";
+
+const MAX_LINES = 2000;
+export interface GetNextResult {
+    pulledLines: number
+}
+
+export function useLogStream(source: string, filterString: string) {
+    const [lines, {clear: clearLines, push: pushLines}] = useList<string>([]);
+    const [cursor, setCursor] = useState<string | null>(null);
+    const [noMoreLinesReason, setNoMoreLinesReason] = useState<string | null>(null);
+    const [hasPulledData, setHasPulledData] = useState(false);
+
+    const hasMore = useMemo(() => {
+        return !hasPulledData || cursor != null;
+    }, [cursor, hasPulledData]);
+
+    useEffect(() => {
+        // if source or filterString changed, clear the state
```

Comment:
This `useEffect` shouldn't be needed if we make `source` and `filterString` as the key of `<ScrollingLogView />` 

e.g.:
```
<ScrollingLogView key={source+filterString} />
```
so as long as those two pieces of info change, the component of `<ScrollingLogView />` is reconstructed. Everything starts clean there is no labor required to maintain a coherent state strenuously. And for such a leaf node, there is not much significance to keep it alive and reuse.

Plus, the `source` and `filterString` can be retrieved from the URL if we intend to take `filterString` into the deep linking mechanism.

For reducing the scope of this PR, I am not including this refactoring in this PR yet. (Maybe later)

Reply:
Sounds good!

Reply:
In another iteration, I found that maintaining a coherent state with `source` and `filterString` without rebuilding `<ScrollingLogView />` increases too much unnecessary code complexity. So I just went ahead and implemented the `keying` strategy mentioned above in this PR.